From 9fdbde80d0aecc82add6abf8a308ca7cd4d351ab Mon Sep 17 00:00:00 2001
From: Ovidiu Panait <ovidiu.panait@windriver.com>
Date: Wed, 23 Oct 2019 15:26:36 +0300
Subject: [PATCH 1/3] Revert "initramfs: don't free a non-existent initrd"

This reverts commit 98d726f8a5d8b6cc272d0c8f6d3995ca98f37d18.

This commit was an adapted version of upstream commit 5d59aa8f9ce9.

However, it doesn't address the initial issue properly. A trace can
still be seen on arm64 during boot with CONFIG_DEBUG_VIRTUAL enabled:

[   44.236944] ------------[ cut here ]------------
[   44.240441] virt_to_phys used for non-linear address: (____ptrval____) (          (null))
[   44.245386] WARNING: CPU: 0 PID: 1 at arch/arm64/mm/physaddr.c:15 __virt_to_phys+0x90/0xc0
[   44.247696] Modules linked in:
[   44.250464] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 4.18.45-yocto-standard #1
[   44.252288] Hardware name: linux,dummy-virt (DT)
[   44.254526] pstate: 80000005 (Nzcv daif -PAN -UAO)
[   44.256409] pc : __virt_to_phys+0x90/0xc0
[   44.258156] lr : __virt_to_phys+0x90/0xc0
[   44.259445] sp : ffffffc06ce5f9f0
[   44.260732] x29: ffffffc06ce5f9f0 x28: ffffffc06ce5fe60
[   44.262934] x27: ffffff90095e31d0 x26: ffffff900957e0a0
[   44.265005] x25: ffffff90095e3000 x24: ffffff90095e2fd0
[   44.267129] x23: ffffff90099a1000 x22: ffffff90099a14c0
[   44.269108] x21: 0000000000000000 x20: 0000000000000000
[   44.271251] x19: 0000000000000000 x18: 000000000000003f
[   44.273179] x17: 0000000000000001 x16: 0000000000000000
[   44.275263] x15: 65701a163a131902 x14: 296c6c756e282020
[   44.277247] x13: 2020202020202020 x12: 2820295f5f5f5f6c
[   44.279338] x11: 61767274705f5f5f x10: 5f28203a73736572
[   44.281416] x9 : 646461207261656e x8 : 696c2d6e6f6e2072
[   44.283433] x7 : 6f66206465737520 x6 : ffffff90099a91dc
[   44.285286] x5 : ffffffc06ce40040 x4 : 0000000000000000
[   44.287270] x3 : dfffff9000000000 x2 : 48342d1a3f2c2800
[   44.289101] x1 : 48342d1a3f2c2800 x0 : 0000000000000000
[   44.291338] Call trace:
[   44.292698]  __virt_to_phys+0x90/0xc0
[   44.294157]  free_initrd_mem+0x64/0x80
[   44.295479]  populate_rootfs+0x208/0x24c
[   44.296776]  do_one_initcall+0x120/0x5cc
[   44.298163]  kernel_init_freeable+0x5b8/0x748
[   44.299457]  kernel_init+0x1c/0x1cc
[   44.300705]  ret_from_fork+0x10/0x18
[   44.302284] ---[ end trace d87d1a81ba85ac56 ]---

Revert this and backport the original commit and a support patch.

Signed-off-by: Ovidiu Panait <ovidiu.panait@windriver.com>
---
 init/initramfs.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/init/initramfs.c b/init/initramfs.c
index 3af2101d19b2..d638443e12ac 100644
--- a/init/initramfs.c
+++ b/init/initramfs.c
@@ -542,7 +542,7 @@ static void __init free_initrd(void)
 	 * If the initrd region is overlapped with crashkernel reserved region,
 	 * free only memory that is not part of crashkernel region.
 	 */
-	if (initrd_start && initrd_start < crashk_end && initrd_end > crashk_start) {
+	if (initrd_start < crashk_end && initrd_end > crashk_start) {
 		/*
 		 * Initialize initrd memory region since the kexec boot does
 		 * not do.
@@ -552,7 +552,7 @@ static void __init free_initrd(void)
 			free_initrd_mem(initrd_start, crashk_start);
 		if (initrd_end > crashk_end)
 			free_initrd_mem(crashk_end, initrd_end);
-	} else if (initrd_start)
+	} else
 #endif
 		free_initrd_mem(initrd_start, initrd_end);
 skip:
-- 
2.17.1

